### PR TITLE
Allow to mark vectors as "userProvided"

### LINKED ISSUE #4606: Avoid regenerating embeddings at dump import
Related product team resources: [PRD](https://www.notion.so/meilisearch/AI-Search-cd5b8d6ce94944618914ab8d958aceb0) (_internal only_)
Related product discussion: https://github.com/orgs/meilisearch/discussions/677

## Motivation

Improve AI search performance

## Usage

Avoid regenerating embeddings at dump import

https://meilisearch.notion.site/v1-9-AI-search-changes-e90d6803eca8417aa70a1ac5d0225697?pvs=74

### Reminders when modifying the Setting API

- [ ] Ensure the new setting route is at least tested by the [`test_setting_routes` macro](https://github.com/meilisearch/meilisearch/blob/5204c0b60b384cbc79621b6b2176fca086069e8e/meilisearch/tests/settings/get_settings.rs#L276)
- [ ] Ensure Analytics are fully implemented
  - [ ] `/settings/my-new-setting` configurated in the [`make_setting_routes` macro](https://github.com/meilisearch/meilisearch/blob/5204c0b60b384cbc79621b6b2176fca086069e8e/meilisearch/src/routes/indexes/settings.rs#L141-L165)
  - [ ] global `/settings` route configurated in the [`update_all` function](https://github.com/meilisearch/meilisearch/blob/5204c0b60b384cbc79621b6b2176fca086069e8e/meilisearch/src/routes/indexes/settings.rs#L655-L751)
- [ ] Ensure the dump serializing is consistent with the `/settings` route serializing, e.g., enums case can be different (`camelCase` in route and `PascalCase` in the dump)

#### Special cases when adding a setting for an experimental feature

- [ ] ⚠️ API stability: The setting does not appear on the main settings route when the feature has never been enabled (e.g. mark it `Unset` when returned from the index in this situation. See [an example](https://github.com/meilisearch/meilisearch/blob/7a89abd2a025606a42f8b219e539117eb2eb029f/meilisearch-types/src/settings.rs#L608))
- [ ] The setting cannot be set when the feature is disabled, either by the main settings route or the subroute (see [`validate_settings` function](https://github.com/meilisearch/meilisearch/blob/7a89abd2a025606a42f8b219e539117eb2eb029f/meilisearch/src/routes/indexes/settings.rs#L811))
- [ ] If possible, the setting is reset when the feature is disabled (hard if it requires reindexing)

## Impacted teams

@meilisearch/docs-team @meilisearch/integration-team

### PR DESCRIPTION
**Pull request by dureuill**
# Pull Request

## Related issue
Fixes #4606 

## What does this PR do?

[See usage in PRD](https://meilisearch.notion.site/v1-9-AI-search-changes-e90d6803eca8417aa70a1ac5d0225697#deb96fb0595947bda7d4a371100326eb)

- Extends the shape of the special `_vectors` field in documents.
    - previously, the `_vectors` field had to be an object, with each field the name of a configured embedder, and each value either `null`, an embedding (array of numbers), or an array of embeddings.
    - In this PR, the value of an embedder in the `_vectors` field can additionally be an object. The object has two fields:
      1. `embeddings`: `null`, an embedding (array of numbers), or an array of embeddings.
      2. `userProvided`: a boolean indicating if the vector was provided by the user.
    - The previous form `embedder_or_array_of_embedders` is semantically equivalent to:
    ```json
    {
        "embeddings": embedder_or_array_of_embedders,
        "userProvided": true
    }
    ```
- During the indexing step, the subfields and values of the `_vectors` field that have `userProvided` set to **false** are added in the vector DB, but not in the documents DB: that means that future modifications of the documents will trigger a regeneration of that particular vector using the document template.
- This allows **importing** embeddings as a one-shot process, while still retaining the ability to regenerate embeddings on document change.
- The dump process now uses this ability: it enriches the `_vectors` fields of documents with the embeddings that were autogenerated, marking them as not `userProvided`. This allows importing the vectors from a dump without regenerating them.

### Tests

This PR adds the following tests

- Long-needed hybrid search tests of a simple hf embedder
- Dump test that imports vectors. Due to the difficulty of actually importing a dump in tests, we just read the dump and check it contains the expected content.
- Tests in the index-scheduler: this tests that documents containing the same kind of instructions as in the dump indexes as expected

### CODE REVIEW

**Comment by dureuill** [comment]
Looks like the vectors in the snapshots as those computed in the CI... Investigating, but we might need a different way to check here.

**Comment by dureuill** [comment]
I made sure the behavior when not passing `embeddings` in a document update (but only `userProvided`): it is the same as passing `embeddings: null`, which results in putting **0** vectors into arroy.

File: milli/src/vector/parsed_vectors.rs [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+        let two_vecs: VectorOrArrayOfVectors =
+            serde_json::from_str("[[0.1, 0.2], [0.3, 0.4]]").unwrap();
+
+        insta::assert_json_snapshot!(null.into_array_of_vectors(), @"null");
+        // 👇 is the the intended behavior? would rather expect [] here, but changing that is a breaking change...
+        insta::assert_json_snapshot!(empty.into_array_of_vectors(), @r###"
+        [
+          []
+        ]
+        "###);
```
Comment (Kerollmops) [inline milli/src/vector/parsed_vectors.rs:171]: Can we change this to return a single `[]`, please?

File: milli/src/update/index_documents/typed_chunk.rs [hunk 2]
```diff
@@ -219,7 +225,34 @@ pub(crate) fn write_typed_chunk_into_index(
                     let del_add_reader = KvReaderDelAdd::new(value);
 
                     if let Some(addition) = del_add_reader.get(DelAdd::Addition) {
-                        writer.insert(field_id, addition)?;
+                        let addition = match vectors_fid {
+                            // for the "_vectors" field, only keep vectors that are marked as userProvided
+                            Some(vectors_fid) if vectors_fid == field_id => 'vectors: {
```
Comment (Kerollmops) [inline milli/src/update/index_documents/typed_chunk.rs:230]:
```suggestion
                        let addition = if vectors_fid == Some(field_id) {
                            // for the "_vectors" field, only keep vectors that are marked as userProvided
                            'vectors: {
```

File: milli/src/vector/parsed_vectors.rs [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        };
+        Ok(ParsedVectors(value))
+    }
+
+    pub fn retain_user_provided_vectors(&mut self) {
+        self.0.retain(|_k, v| match v {
+            Vectors::ImplicitlyUserProvided(_) => true,
+            Vectors::Explicit(ExplicitVectors { embeddings: _, user_provided }) => *user_provided,
+        });
+    }
```
Comment (Kerollmops) [inline milli/src/vector/parsed_vectors.rs:97]: This function should keep vectors associated with an embedder that does not exist.

**Review by Kerollmops — CHANGES_REQUESTED** [review]

File: milli/src/error.rs [hunk 4]
```diff
@@ -120,7 +120,7 @@ only composed of alphanumeric characters (a-z A-Z 0-9), hyphens (-) and undersco
     #[error("The `_vectors.{subfield}` field in the document with id: `{document_id}` is not an array. Was expecting an array of floats or an array of arrays of floats but instead got `{value}`.")]
     InvalidVectorsType { document_id: Value, value: Value, subfield: String },
```
[the remark this answers is not in the document]
Comment (dureuill) [reply milli/src/error.rs]: Nope, removed

File: milli/src/update/index_documents/extract/extract_vector_points.rs [hunk 5]
```diff
@@ -114,152 +151,140 @@ pub fn extract_vector_points<R: io::Read + io::Seek>(
...
-    Ok(ExtractedVectorPoints {
-        // docid, _index -> KvWriterDelAdd -> Vector
-        manual_vectors: writer_into_reader(manual_vectors_writer)?,
-        // docid -> ()
-        remove_vectors: writer_into_reader(remove_vectors_writer)?,
-        // docid -> prompt
-        prompts: writer_into_reader(prompts_writer)?,
-    })
-}
+    /////
```
[the remark this answers is not in the document]
Comment (dureuill) [reply milli/src/update/index_documents/extract/extract_vector_points.rs]: Removed

File: milli/src/update/index_documents/extract/extract_vector_points.rs [hunk 6]
```diff
@@ -65,42 +69,75 @@ impl VectorStateDelta {
...
 /// Returns the generated grenad reader containing the docid as key associated to the Vec<f32>
 #[tracing::instrument(level = "trace", skip_all, target = "indexing::extract")]
 pub fn extract_vector_points<R: io::Read + io::Seek>(
     obkv_documents: grenad::Reader<R>,
     indexer: GrenadParameters,
-    settings_diff: &InnerIndexSettingsDiff,
-    prompt: &Prompt,
-    embedder_name: &str,
-) -> Result<ExtractedVectorPoints> {
+    settings_diff: Arc<InnerIndexSettingsDiff>,
```
[the remark this answers is not in the document]
Comment (dureuill) [reply milli/src/update/index_documents/extract/extract_vector_points.rs]: done

File: milli/src/vector/parsed_vectors.rs (hunk 3 above)
Comment (dureuill) [reply milli/src/vector/parsed_vectors.rs]: Changed to do so

File: milli/src/update/index_documents/typed_chunk.rs (hunk 2 above)
Comment (dureuill) [reply milli/src/update/index_documents/typed_chunk.rs]: changed

File: milli/src/error.rs [hunk 7]
```diff
@@ -120,7 +120,7 @@ only composed of alphanumeric characters (a-z A-Z 0-9), hyphens (-) and undersco
...
     #[error("The `_vectors` field in the document with id: `{document_id}` is not an object. Was expecting an object with a key for each embedder with manually provided vectors, but instead got `{value}`")]
-    InvalidVectorsMapType { document_id: Value, value: Value },
```
[the remark this answers is not in the document]
Comment (dureuill) [reply milli/src/error.rs:123]: On second thought it looks like the error message is still current, if a bit imprecise.

File: milli/src/vector/parsed_vectors.rs (hunk 1 above)
Comment (dureuill) [reply milli/src/vector/parsed_vectors.rs]: done

**Review by Kerollmops — DISMISSED** [review]
Thank you very much @dureuill. LGTM 🍸
bors merge

**Comment by Kerollmops** [comment]
bors cancel

**Comment by Kerollmops** [comment]
bors merge

**Review by Kerollmops — APPROVED** [review]
bors merge

**Comment by Kerollmops** [comment]
bors merge

**Comment by dureuill** [comment]
bors merge

**Pull request merged**